### PR TITLE
[BC] Change how resource pack grouping works and group mod data packs

### PR DESCRIPTION
**Pull request by lukebemish**
Currently, NeoForge groups resource packs from various mods into a single resource pack via `DelegatingPackResources`, a PackResources which groups others together. Doing so requires adding the `getChildren` extension to `PackResources`, and changes the nature of `PackResources` from "a single member of a resource stack" to "one or more layers of a resource stack". This is, in most situations, not an issue.

However, vanilla has added more and more features that are controllable by the pack.mcmeta file - and accessed through the `getMetadataSection` of `PackResources`, including overlays and filtering. Data and resource packs make use of these features; mods may as well. The use of `DelegatingPackResources` to group mod resources means that this information is lost, and mods (as well as resource packs packaged as low-code mods) cannot make use of these features. Additionally, the changed behaviour of PackResources can easily lead to confusion while using them, as I have experienced personally - though this particular issue could be mitigated by some javadoc changes.

While for resource packs these limits are rather benign ("could cause issues" not "has caused issues"), as the features in question aren't used by many resource packs and low-code resource packs are rare, they become rather more problematic if, in the future, we wished to group mod *data*, not just assets - which is probably a good idea to do at some point. Low-code datapacks are much more common, and filters and overlays are much more useful in datapacks.

My proposal here is to switch the grouping from `PackResources` to `Pack`; each `Pack` may now optionally hold a number of "child" packs (which are required to be marked as hidden). When the selected pack list is being rebuilt in `PackRepository`, these child packs are injected, in order, directly beneath the parent pack; this means that they will stay with the parent pack, even as packs are re-ordered, shuffled around, whatever! From the user perspective, only the parent pack is visible (if it isn't hidden). From a code perspective, listing selected packs from a `PackRepository` and onwards, *all* the packs in question are visible, with the child packs always directly below the parent pack, in order.

----------

This PR currently switches NeoForge's mod asset loading to use grouped packs instead of grouped pack resources, and replaces `DelegatingPackResources` with `EmptyPackResources`, a `PackResources` that has metadata and nothing else. I would welcome feedback on the approach in general, but am also looking specifically for good test cases to make sure I didn't frick anything up.

Beyond changing resource pack grouping, this PR uses the new grouping system to consolidate logic for mod data and resource pack setup - which includes grouping mod datapacks together as well. This should ensure that mod datapacks are always placed in mod load order, not "order I installed the mods in" order as they currently are, which should help solve some of the funky issues that can pop up there.

### CODE REVIEW

File: patches/minecraft/net/minecraft/server/packs/repository/Pack.java.patch [hunk 1]
```diff
@@ -1,21 +1,40 @@
...
++      builder.addAll(this.children);
++      builder.addAll(children);
++      // data and asset formats passed in pack info here are irrelevant, as they are not used in Pack constructor
++      return new Pack(this.f_10401_, this.f_10407_, this.f_244124_, this.f_10403_, new Pack.Info(this.f_10404_, 0, 0, this.f_244623_, this.hidden), this.f_10405_, this.f_10406_, this.f_10408_, this.f_10409_, builder.build());
++   }
++
++   private Pack(String p_252218_, boolean p_248829_, Pack.ResourcesSupplier p_249377_, Component p_251718_, Pack.Info p_250162_, PackCompatibility p_250361_, Pack.Position p_251298_, boolean p_249753_, PackSource p_251608_, List<Pack> children) {
++      for (Pack child : children) {
++         if (!child.isHidden()) {
++            throw new IllegalArgumentException("Child pack must be hidden");
```
Comment (XFactHD) [inline patches/minecraft/net/minecraft/server/packs/repository/Pack.java.patch]: The pack ID of the non-hidden child pack should be included in the exception message

File: patches/minecraft/net/minecraft/server/packs/repository/PackRepository.java.patch [hunk 2]
```diff
@@ -9,6 +9,23 @@
...
+ 
+    private List<Pack> m_10517_(Collection<String> p_10518_) {
+-      List<Pack> list = this.m_10520_(p_10518_).collect(Collectors.toList());
++      List<Pack> list = this.m_10520_(p_10518_).flatMap(p -> Stream.concat(Stream.of(p), p.getChildren().stream())).collect(Collectors.toList());
+ 
+       for(Pack pack : this.f_10498_.values()) {
+          if (pack.m_10449_() && !list.contains(pack)) {
+-            pack.m_10451_().m_10470_(list, pack, Functions.identity(), false);
++            int i = pack.m_10451_().m_10470_(list, pack, Functions.identity(), false);
++            for (int j = 0; j < pack.getChildren().size(); j++) {
```
Comment (XFactHD) [inline patches/minecraft/net/minecraft/server/packs/repository/PackRepository.java.patch]: I would use `list.addAll(i + 1, pack.getChildren());` instead of `add()` in a loop.

File: src/main/java/net/minecraftforge/client/loading/ClientModLoader.java [hunk 3]
```diff
@@ -179,11 +180,12 @@ private static void clientPackFinder(Map<IModFile, ? extends PathPackResources>
             }
         }
 
-        // Create a resource pack merging all mod resources that should be hidden
-        final Pack modResourcesPack = Pack.readMetaAndCreate("mod_resources", Component.literal("Mod Resources"), true,
-                id -> new DelegatingPackResources(id, false, new PackMetadataSection(Component.translatable("fml.resources.modresources", hiddenPacks.size()),
-                        SharedConstants.getCurrentVersion().getPackVersion(PackType.CLIENT_RESOURCES)), hiddenPacks),
-                PackType.CLIENT_RESOURCES, Pack.Position.BOTTOM, PackSource.DEFAULT);
-        packAcceptor.accept(modResourcesPack);
+        final Pack modMarkerPack = Pack.readMetaAndCreate("mod_resources_marker", Component.literal("Mod Resources"), true,
```
Comment (XFactHD) [inline src/main/java/net/minecraftforge/client/loading/ClientModLoader.java]: This needs some linebreaks, as-is it's almost impossible follow what this nested and chained statement does

File: src/main/java/net/minecraftforge/resource/ResourcePackLoader.java [hunk 4]
```diff
@@ -63,6 +64,11 @@ protected Path resolve(@NotNull String... paths)
             {
                 return this.modFile.findResource(paths);
             }
+
+            @Override
+            public boolean isHidden() {
+                return true;
```
Comment (XFactHD) [inline src/main/java/net/minecraftforge/resource/ResourcePackLoader.java]: This codepath also creates the `PackResources` for mods which request to be shown as a distinct pack in the pack list screen, so this must return `!mf.showAsResourcePack()` instead of `true`.

File: src/main/java/net/minecraftforge/resource/ResourcePackLoader.java [hunk 5]
```diff
@@ -53,7 +53,8 @@ public static void loadResourcePacks(PackRepository resourcePacks, Function<Map<
     @NotNull
     public static PathPackResources createPackForMod(IModFileInfo mf)
     {
-        return new PathPackResources(mf.getFile().getFileName(), true, mf.getFile().getFilePath())
+        String name = mf.getFile().getFileName();
+        return new PathPackResources(name.isEmpty() ? "unnamed" : name, true, mf.getFile().getFilePath())
```
Comment (XFactHD) [inline src/main/java/net/minecraftforge/resource/ResourcePackLoader.java]: I think making the `"unnamed"` name a bit more distinct would be good (i.e. wrapping it in square brackets like `"[unnamed]"`) since there are cases where a non-filename name can be shown here, specifically in a dev environment.

**Review by XFactHD — CHANGES_REQUESTED** [review]
Overall this looks significantly cleaner in practice than in my mind. I quickly threw a build of this at AtlasViewer together with JEI to have another mod with resources in the environment and my "source pack" collection worked flawlessly with this and became simpler as well due to not needing a special case for the delegating pack (which I completely forgot existed when this was discussed on Discord).

I've noted a few issues and a few minor changes.

**Comment by lukebemish** [comment]
~~Marking this PR as no longer a draft as I think code wise it's in a place where more review is welcome, but note that I've still only tested this in a couple of cases and haven't figured out a good list of weird edge cases to throw at it - and the datapack side of stuff in particular could do with more testing!~~ scratch that, I'll wait to mark it ready for review until I've got a PR to forgespi open to remove those remaining "TODO" blocks - basically, give mods separate ways to mark themselves as wanting a separately visible data/resource pack

**Comment by lukebemish** [comment]
PRs to NeoForgeSPI and FancyModLoader opened for that needed flag for datapack grouping; see NeoForged/FancyModLoader#21 and NeoForged/NeoForgeSPI#1

**Comment by KnightMiner** [comment]
Would it be feasible to make this change without it being a breaking change by just deprecating the old method and keeping the delegating resource pack around but not used? No one should be relying on the fact that mod resources are delegating, rather they should just iterate though the list of all packs and use the forge added method as needed.

**Comment by lukebemish** [comment]
The resource grouping part? Possibly, though I would argue it's better to roll it into one big change - a big part of the point of this after all is to change back the assumption of PackResources to the vanilla "one resource per PackResource", which then simplifies a ton of other stuff. The grouping mod datapack part? Definitely not, that requires breaking changes to NeoForgeSPI

File: src/main/java/net/minecraftforge/resource/ResourcePackLoader.java [hunk 6]
```diff
@@ -6,54 +6,97 @@
...
+            if (Objects.equals(mod.getModId(), "minecraft")) continue; // skip the minecraft "mod"
+            final String name = "mod:" + mod.getModId();
+            final Pack modPack = Pack.readMetaAndCreate(name, Component.literal(e.getValue().packId()), false, id -> e.getValue(), packType, Pack.Position.BOTTOM, PackSource.DEFAULT);
+            if (modPack == null) {
+                // Vanilla only logs an error, instead of propagating, so handle null and warn that something went wrong
+                ModLoader.get().addWarning(new ModLoadingWarning(mod, ModLoadingStage.ERROR, "fml.modloading.brokenresources", e.getKey()));
+                continue;
+            }
+            LOGGER.debug(Logging.CORE, "Generating PackInfo named {} for mod file {}", name, e.getKey().getFilePath());
+            if ((packType == PackType.CLIENT_RESOURCES && mod.getOwningFile().showAsResourcePack()) || mod.getOwningFile().showAsDataPack()) {
```
Comment (XFactHD) [inline src/main/java/net/minecraftforge/resource/ResourcePackLoader.java]: This doesn't do what you want it to do, it will incorrectly add a resource pack as standalone if the mod says it wants a standalone datapack. You need to either check `packType == PackType.CLIENT_RESOURCES ? mod.getOwningFile().showAsResourcePack() : mod.getOwningFile().showAsDataPack()` or `(packType == PackType.CLIENT_RESOURCES && mod.getOwningFile().showAsResourcePack()) || (packType == PackType.SERVER_DATA && mod.getOwningFile().showAsDataPack())`. The same applies to the `PackResources#isHidden()` implementation in `createPackForMod()`

Comment (lukebemish) [reply src/main/java/net/minecraftforge/resource/ResourcePackLoader.java]: Should be fixed now

**Comment by lukebemish** [comment]
> The resource grouping part? Possibly, though I would argue it's better to roll it into one big change - a big part of the point of this after all is to change back the assumption of PackResources to the vanilla "one resource per PackResource", which then simplifies a ton of other stuff. The grouping mod datapack part? Definitely not, that requires breaking changes to NeoForgeSPI

To go off of this - machermans pointed out that the spi changes could be made non breaking, so yes, it is technically possible. However, I would argue that it is still not worth it, given that one of the major goals of this PR is removing the concept of nested pack resources entirely

**Comment by lukebemish** [comment]
Superseded by #367
